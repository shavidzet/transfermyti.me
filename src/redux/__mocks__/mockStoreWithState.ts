import { ActionCreator, Middleware } from "redux";
import configureStore, { MockStore } from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { default as stateFixture } from "../__fixtures__/state";

export { MockStoreCreator, MockStore } from "redux-mock-store";

// This is the same as ReduxAction, but the type was changed to `any`, otherwise
// we'd have to explicitly specify every action `type` name when using jest-in-case:
export interface ReduxTestAction<TPayload = {}> {
  type: any;
  payload?: TPayload;
  error?: boolean;
  meta?: any;
}

export type ReduxTestActionCreator<TPayload> = ActionCreator<
  ReduxTestAction<TPayload>
  >;

// This is done to simplify the imports in tests (rather than have to import
// state/fixtures from the state.ts file and mockStoreWithState() function from
// another:
export const state = stateFixture;

/**
 * Creates an instance of a Redux store for mocking in tests.
 * @param stateChanges Updated entries in state (doesn't have to be entire state
 *                     since state fixture is used.
 * @param additionalMiddleware Any other middlewares to pass to configureStore
 *                             function.
 */
export function mockStoreWithState<TState>(
  stateChanges?: TState,
  additionalMiddleware: Array<Middleware> = [],
): MockStore {
  const sagaMiddleware = createSagaMiddleware();
  const mockStore = configureStore([sagaMiddleware, ...additionalMiddleware]);
  const validState = { ...stateFixture, ...stateChanges };
  return mockStore(validState);
}
