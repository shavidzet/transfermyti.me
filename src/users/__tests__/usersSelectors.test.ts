import cases from "jest-in-case";
import { state } from "~/redux/__mocks__/mockStoreWithState";
import * as usersSelectors from "../usersSelectors";

const TEST_STATE = {
  ...state,
  users: {
    source: {
      "6001": {
        id: "6001",
        name: "John Test",
        email: "test-user@test.com",
        isAdmin: null,
        isActive: true,
        userGroupIds: [],
        workspaceId: "1001",
        entryCount: 0,
        linkedId: "clock-user-01",
        isIncluded: false,
        memberOf: "users",
      },
      "6002": {
        id: "6002",
        name: "Mary Testarossa",
        email: "mary@testarossa.com",
        isAdmin: null,
        isActive: true,
        userGroupIds: [],
        workspaceId: null,
        entryCount: 0,
        linkedId: null,
        isIncluded: true,
        memberOf: "users",
      },
    },
    target: {
      "clock-user-01": {
        id: "clock-user-01",
        name: "John Test",
        email: "test-user@test.com",
        isAdmin: null,
        isActive: true,
        userGroupIds: [],
        workspaceId: "clock-workspace-01",
        entryCount: 0,
        linkedId: "6001",
        isIncluded: false,
        memberOf: "users",
      },
    },
    isFetching: false,
  },
};

describe("within usersSelectors", () => {
  test("the sourceUsersSelector returns array of values from state.source", () => {
    const result = usersSelectors.sourceUsersSelector(TEST_STATE);

    expect(result).toEqual(Object.values(TEST_STATE.users.source));
  });

  cases(
    "the selectors match their snapshots",
    options => {
      const result = options.selector(TEST_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceUsersSelector",
        selector: usersSelectors.includedSourceUsersSelector,
      },
      {
        name: "for the userIdToLinkedIdSelector",
        selector: usersSelectors.userIdToLinkedIdSelector,
      },
      {
        name: "for the sourceUsersForTransferSelector",
        selector: usersSelectors.sourceUsersForTransferSelector,
      },
      {
        name: "for the sourceUserEmailsByWorkspaceIdSelector",
        selector: usersSelectors.sourceUserEmailsByWorkspaceIdSelector,
      },
    ],
  );
});
