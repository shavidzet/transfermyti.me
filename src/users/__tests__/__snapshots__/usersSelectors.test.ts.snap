// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within usersSelectors the selectors match their snapshots for the includedSourceUsersSelector 1`] = `
Array [
  Object {
    "email": "mary@testarossa.com",
    "entryCount": 0,
    "id": "6002",
    "isActive": true,
    "isAdmin": null,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "users",
    "name": "Mary Testarossa",
    "userGroupIds": Array [],
    "workspaceId": null,
  },
]
`;

exports[`within usersSelectors the selectors match their snapshots for the sourceUserEmailsByWorkspaceIdSelector 1`] = `
Object {
  "clock-workspace-01": Array [
    "test-user@test.com",
  ],
}
`;

exports[`within usersSelectors the selectors match their snapshots for the sourceUsersForTransferSelector 1`] = `
Array [
  Object {
    "email": "mary@testarossa.com",
    "entryCount": 0,
    "id": "6002",
    "isActive": true,
    "isAdmin": null,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "users",
    "name": "Mary Testarossa",
    "userGroupIds": Array [],
    "workspaceId": null,
  },
]
`;

exports[`within usersSelectors the selectors match their snapshots for the userIdToLinkedIdSelector 1`] = `
Object {
  "6001": "clock-user-01",
  "clock-user-01": "6001",
}
`;
